src/solver/infeasible-problem-analysis/watched-constraints.cpp

```diff
@@ -0,0 +1,217 @@
+           + timeStep(splitName());
+}
+
+std::string HydroLevel::infeasibilityCause()
+{
+    return "* Hydro reservoir impossible to manage with cumulative options "
+           "\"hard bounds without heuristic\"";
+}
+
+// --- Short term storage constraint ---
+std::string STS::infeasibility()
+{
+    return "Short-term-storage reservoir constraint at area '" + areaName(splitName())
+           + "' in STS '" + STSname(splitName()) + "' at hour " + timeStep(splitName());
+}
+
+std::string STS::infeasibilityCause()
+{
+    return "* Short-term storage reservoir level impossible to manage. Please check inflows, "
+           "lower & upper curves and initial level (if prescribed),";
+}
+
+// --- Hydro production constraint ---
+std::string HydroProduction::infeasibility()
+{
+    return "Hydro weekly production at area '" + areaName(splitName()) + "'";
+}
+
+std::string HydroProduction::infeasibilityCause()
+{
+    return "* impossible to generate exactly the weekly hydro target";
+}
+
+// --- Constraints factory ---
+std::shared_ptr<WatchedConstraint> createHourlyBC(std::string varName)
+{
+    return std::make_shared<HourlyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createDailyBC(std::string varName)
+{
+    return std::make_shared<DailyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createWeeklyBC(std::string varName)
+{
+    return std::make_shared<WeeklyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createFictitiousLoad(std::string varName)
+{
+    return std::make_shared<FictitiousLoad>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createHydroLevel(std::string varName)
+{
+    return std::make_shared<HydroLevel>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createSTS(std::string varName)
+{
+    return std::make_shared<STS>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createHydroProduction(std::string varName)
+{
+    return std::make_shared<HydroProduction>(varName);
+}
+
+ConstraintsFactory::ConstraintsFactory()
+{
+    regex_to_ctypes_["::hourly::"] = createHourlyBC;
+    regex_to_ctypes_["::daily::"] = createDailyBC;
+    regex_to_ctypes_["::weekly::"] = createWeeklyBC;
+    regex_to_ctypes_["^FictiveLoads::"] = createFictitiousLoad;
+    regex_to_ctypes_["^AreaHydroLevel::"] = createHydroLevel;
+    regex_to_ctypes_["^Level::"] = createSTS;
+    regex_to_ctypes_["^HydroPower::"] = createHydroProduction;
+
+    auto keyView = std::views::keys(regex_to_ctypes_);
+    regex_ids_ = {keyView.begin(), keyView.end()};
+}
+
+std::shared_ptr<WatchedConstraint> ConstraintsFactory::create(std::string varName) const
+{
+    return std::ranges::find_if(regex_to_ctypes_,
+                                [&varName](auto& p)
+                                { return std::regex_search(varName, std::regex(p.first)); })
+      ->second(varName);
+}
+
+std::regex ConstraintsFactory::constraintsFilter()
+{
+    return std::regex(boost::algorithm::join(regex_ids_, "|"));
```

Comment:
```suggestion
    return std::regex(boost::algorithm::join(regex_to_ctypes_ | std::views::keys, "|"));
```
does this work ? If so, `regex_ids_` can be removed.

Reply:
Sorry I couldn't make it compile like that